crates/bootstrap_srv/src/test.rs

```diff
@@ -0,0 +1,773 @@
+        ..Default::default()
+    }
+    .call()
+    .unwrap_err();
+
+    assert!(err.to_string().contains("InvalidSignature"));
+}
+
+#[test]
+fn reject_bad_agent_pub_key() {
+    let s = BootstrapSrv::new(Config::testing()).unwrap();
+
+    let err = PutInfo {
+        addr: s.listen_addr(),
+        // only 31 characters... and obviously the wrong key : )
+        final_agent_pk: Some("AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"),
+        ..Default::default()
+    }
+    .call()
+    .unwrap_err();
+
+    assert!(err.to_string().contains("InvalidAgentPubKey"));
+}
+
+#[test]
+fn default_storage_rollover() {
+    let s = BootstrapSrv::new(Config::testing()).unwrap();
+
+    let addr = s.listen_addr();
+    let mut test_prop: u32 = 0;
+    let mut put_info = move || {
+        use base64::prelude::*;
+        let mut agent_seed = [0; 32];
+        agent_seed[..4].copy_from_slice(&test_prop.to_le_bytes());
+        let agent_seed = BASE64_URL_SAFE_NO_PAD.encode(agent_seed);
+        PutInfo {
+            addr,
+            agent_seed: &agent_seed,
+            test_prop: &format!("{test_prop}"),
+            ..Default::default()
+        }
+        .call()
+        .unwrap();
+        test_prop += 1;
+    };
+
+    let addr = s.listen_addr();
+    let get = move || {
+        let addr = format!("http://{:?}/bootstrap/{}", addr, S1);
+        let res = ureq::get(&addr).call().unwrap().into_string().unwrap();
+        let res: Vec<DecodeAgent> = serde_json::from_str(&res).unwrap();
+        res.into_iter().map(|m| m.test_prop).collect::<Vec<_>>()
+    };
+
+    for _ in 0..32 {
+        put_info();
+    }
+
+    let res = get();
+
+    assert_eq!(
+        res.as_slice(),
+        &[
+            "0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12",
+            "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "23",
+            "24", "25", "26", "27", "28", "29", "30", "31",
+        ]
+    );
+
+    for _ in 0..32 {
+        put_info();
+    }
+
+    let res = get();
+
+    // Agents put beyond MAX_STORAGE will be cause the item at
+    // MAX_STORAGE  / 2 to be deleted, then the new put will be appended
+    // to the end. The deleted item here is 32 / 2 = index 16, so index 17
+    // onward will have the higher test_prop values as they rolled over.
+    assert_eq!(
+        res.as_slice(),
+        &[
+            "0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12",
+            "13", "14", "15", "48", "49", "50", "51", "52", "53", "54", "55",
+            "56", "57", "58", "59", "60", "61", "62", "63",
+        ]
+    );
+}
+
+#[test]
+fn multi_thread_stress() {
+    let s = BootstrapSrv::new(Config::testing()).unwrap();
+    let addr = s.listen_addr();
+
+    let start = std::time::Instant::now();
+
+    // the testing config has 2 worker threads.
+    // Read and write with more than that.
+    const T_W_COUNT: u32 = 16;
```

Comment:
We probably won't change these (ever?), but if we did, would it be helpful to make these multiples of the configured number of workers in the testing config? I.e. `let t_w_count = config.worker_thread_count * 8`.

Reply:
I'm not sure there's any particular reason to have an even multiple of the worker threads. It should work totally fine to have it be any value.

Reply:
No, I misphrased that. Instead of a hardcoded `const` use a calculated number with `config.worker_thread_count`, e.g. `let t_w_count = config.worker_thread_count + x`.

Reply:
Ah, I understand : ) Done: https://github.com/holochain/kitsune2/pull/11/commits/826d7de0f4cdaaab5125ce0000f166198ef2aa41

Reply:
One more suggestion:
```rust
let config = Config::testing();
let s = BootstrapSrv::new(config.clone()).unwrap();
let addr = s.listen_addr();

let start = std::time::Instant::now();

// the testing config has a small number worker threads (currently 2).
// Read and write with more than that.
let worker_count = config.worker_thread_count as u32;
```

Reply:
Done: https://github.com/holochain/kitsune2/pull/11/commits/d803ec7e3599d8d7900f2bccce0171303a6ce6ea

Reply:
Thanks for your patience here =)